src/ui/datacore-view-page.tsx

```diff
@@ -0,0 +1,241 @@
+                    onValueChange={(s) => setInternalState("currentFile", s)}
+                />
+            </Group>
+        </Stack>
+    );
+}
+
+export interface DatacoreViewState {
+    sourceType: /* "datacore" | */ ScriptLanguage;
+    script: string;
+    currentFile: string;
+    title: string;
+}
+
+export class DatacoreQueryView extends ItemView {
+    public internalState: DatacoreViewState = {
+        title: "",
+        currentFile: "",
+        script: "",
+        sourceType: "js",
+    };
+    public jsRenderer: DatacoreJSRenderer;
+    public settingsRenderer: ReactRenderer;
+    public isSettingsShowing: boolean;
+    private id: string;
+    navigation: boolean = false;
+    constructor(leaf: WorkspaceLeaf, public api: DatacoreApi, private plugin: DatacorePlugin) {
+        super(leaf);
+        this.rerender();
+    }
+    getViewType(): string {
+        return VIEW_TYPE_DATACORE;
+    }
+    getDisplayText(): string {
+        return `${this.internalState.title} (Datacore)`;
+    }
+    public async onload() {
+        this.settingsRenderer = new ReactRenderer(
+            this.app,
+            this.api.core,
+            this.contentEl,
+            this.internalState.currentFile || "",
+            (
+                <CUSTOM_VIEW_CONTEXT.Provider value={this}>
+                    <DatacoreViewSettings />
+                </CUSTOM_VIEW_CONTEXT.Provider>
+            )
+        );
+
+        this.id = this.leaf.serialize().id;
+				this.contentEl.addClass("markdown-rendered");
+        this.rerender();
+    }
+    public onPaneMenu(menu: Menu, source: "more-options" | "tab-header" | string): void {
+        if (source === "more-options") {
+            menu.addItem((it) => {
+                it.setIcon("settings");
+                it.setTitle("View configuration");
+                it.onClick((e) => {
+                    this.toggleSettings(true);
+                });
+            });
+        }
+    }
+    private settingsShowing: boolean;
+    public toggleSettings(show: boolean): void {
+        this.settingsShowing = show;
+        if (show) {
+            this.settingsRenderer = new ReactRenderer(
+                this.app,
+                this.api.core,
+                this.contentEl,
+                this.internalState.currentFile || "",
+                (
+                    <CUSTOM_VIEW_CONTEXT.Provider value={this}>
+                        <DatacoreViewSettings />
+                    </CUSTOM_VIEW_CONTEXT.Provider>
+                )
+            );
+            this.removeChild(this.jsRenderer);
+            this.addChild(this.settingsRenderer);
+        } else {
+            this.removeChild(this.settingsRenderer);
+            this.rerender();
+        }
+    }
+    public getState() {
+        return this.internalState;
+        // return this.internalState;
+    }
+    public async onOpen(): Promise<void> {}
+    rerender(): void {
+        this.removeChild(this.jsRenderer);

```

Comment:
jsRenderer will be null on the first invocation here, is that ok?

Reply:
i've now added a null check.